### PR TITLE
Add opt-in draft vocab padding for speculative decoding

### PR DESCRIPTION
**Pull request by massif-01**
## Purpose

This PR adds an explicit opt-in path for draft model vocabulary padding in speculative decoding.

The motivating case is using a smaller Qwen-family draft model with a larger Qwen-family target model where the token IDs are suffix-compatible: the draft model uses the same prefix token IDs, while the target model only adds extra tokens at the tail of the vocabulary.

In my local deployment, I hit this with:

- target model: `DeepSeek-R1-Distill-Qwen-32B-FP8`
- draft model: `DeepSeek-R1-Distill-Qwen-1.5B`

The target and draft vocab sizes differ, so vLLM's current strict speculative decoding checks reject the pair. My first local workaround was to manually edit the draft model `config.json` to the target vocab size, then loosen the vocab embedding loader so a shorter checkpoint weight could be loaded and the missing tail rows would remain zero-filled.

That workaround was enough to make the workload start, but it was too broad for upstream: it made normal model loading accept shorter vocab weights, which could hide real checkpoint/config mismatches, tokenizer incompatibilities, or packed/quantized weight issues. That would be risky for ordinary models and newer Qwen-family models such as Qwen3/Qwen3.5.

This PR narrows the behavior into a guarded opt-in feature:

- adds `SpeculativeConfig.allow_draft_model_vocab_padding`, defaulting to `False`;
- only allows the option for `method == "draft_model"`;
- only allows it when the target vocab size is larger than the draft vocab size;
- records the original draft checkpoint vocab size as `draft_model_unpadded_vocab_size`;
- updates the draft runtime vocab size to the target vocab size, so users do not need to edit the draft model config file manually;
- keeps normal model loading and normal speculative decoding strict by default;
- only allows short-vocab loading for draft runner vocab weights when the loaded checkpoint vocab exactly matches the recorded original draft vocab size;
- rejects unsupported cases such as packed vocab weights, non-`output_dim=0` short loading, and added vocabulary embeddings;
- masks the zero-padded token range in both full-logits and local-argmax draft selection paths so the draft model cannot select synthetic zero-filled lm_head rows.

This is intentionally not a general heterogeneous-tokenizer solution. It only supports suffix-compatible vocab expansion where token IDs already match for the draft vocab prefix. General tokenizer/vocab remapping is a broader problem and is being addressed by the TLI direction in #38174. I checked current open PRs and found no duplicate PR for this narrower opt-in padding path; #38174 targets universal heterogeneous vocabulary speculative decoding, while this PR keeps the existing `draft_model` path strict by default and only adds a small opt-in escape hatch for suffix-compatible vocab expansion.

AI assistance was used to help prepare and validate this change. The human submitter reviewed the changed code and test results.

The local workload that motivated this change was started with:

```bash
vllm serve "/home/rm01/Downloads/deepseek-ai/DeepSeek-R1-Distill-Qwen-32B-FP8" \
  --gpu-memory-utilization=0.88 \
  --max-model-len=32000 \
  --served-model-name "DeepSeek-R1-Distill-Qwen-32B-Speculative-1.5B-32K" \
  --enable-prefix-caching \
  --enable-chunked-prefill \
  --speculative-model "/home/rm01/Downloads/deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B" \
  --num-speculative-token 5 \
  --speculative-draft-tensor-parallel-size 1 \
  --max_num_batched_tokens 256
```

With this PR, the new opt-in field must be enabled in the speculative config for this suffix-compatible short-draft-vocab case.

## Test Plan

Added targeted tests for:

- default strict behavior when a shorter vocab weight is loaded without opt-in;
- default target/draft vocab mismatch rejection;
- `allow_draft_model_vocab_padding=True` validation for `draft_model` only;
- target vocab size must be larger than draft vocab size;
- draft runtime vocab config update while preserving the original draft vocab metadata;
- TP=1/2/4 short-vocab weight loading with shard-overlap copying and zero-filled missing rows;
- a rank whose shard is fully in the missing suffix remains zero-filled;
- loaded vocab sizes shorter or longer than the recorded original draft vocab size are rejected;
- packed weights and non-`output_dim=0` short loading are rejected;
- full logits mask the padded token range to `-inf`;
- local argmax cannot select zero-filled padded lm_head rows.

Validation commands:

```bash
git diff --check
```

```bash
PYTHONPATH=. .venv/bin/python -m pytest \
  tests/config/test_speculative_vocab_padding.py \
  tests/model_executor/layers/test_vocab_parallel_embedding.py \
  -q
```

```bash
PRE_COMMIT_HOME=/private/tmp/codex-pre-commit .venv/bin/pre-commit run --files \
  vllm/config/speculative.py \
  vllm/model_executor/layers/vocab_parallel_embedding.py \
  vllm/model_executor/layers/logits_processor.py \
  tests/config/test_speculative_vocab_padding.py \
  tests/model_executor/layers/test_vocab_parallel_embedding.py
```

## Test Result

`git diff --check`:

```text
Passed
```

Targeted unit tests:

```text
21 passed, 17 warnings in 0.48s
```

Additional independent validation was provided by @jmamou in
https://github.com/vllm-project/vllm/pull/42775#issuecomment-4544136751.

They tested the same padded-vocab pattern with:

- Target: `Qwen/Qwen2.5-Coder-14B-Instruct` (`vocab_size=152064`)
- Draft: `Qwen/Qwen2.5-Coder-0.5B-Instruct` (`vocab_size=151936`)

The two tokenizers share the same 151,665 actual tokens, and the vocab-size
difference is only tail embedding-row padding. With
`allow_draft_model_vocab_padding=true`, speculative decoding ran cleanly on a
SPEED-Bench `low_entropy.code_completion` workload with no correctness issues
observed across the sweep.

Reported TPOT speedups vs baseline:

| BS | k | TPOT (ms) | speedup |
|---|---|---:|---:|
| 1 | 7 | 110 | 2.82x |
| 4 | 7 | 348 | 1.98x |
| 8 | 5 | 628 | 1.43x |
| 8 | 7 | 609 | 1.48x |

This supports the intended scope of this PR: a narrow opt-in for
same-tokenizer-family models with prefix-compatible token IDs and padded target
vocab tails, while leaving the broader heterogeneous-tokenizer case to #38174.

Pre-commit on changed files:

```text
ruff check..........................................................................................Passed
ruff format.........................................................................................Passed
typos...............................................................................................Passed
clang-format....................................................................(no files to check)Skipped
markdownlint-cli2...............................................................(no files to check)Skipped
Lint GitHub Actions workflow files..............................................(no files to check)Skipped
pip-compile.....................................................................(no files to check)Skipped
pip-compile-rocm................................................................(no files to check)Skipped
pip-compile-xpu.................................................................(no files to check)Skipped
pip-compile-docs................................................................(no files to check)Skipped
reformat test/nightly-torch.txt to be in sync with test/cuda.in.................(no files to check)Skipped
Run mypy locally for lowest supported Python version............................Passed
Lint shell scripts..............................................................(no files to check)Skipped
Lint PNG exports from excalidraw................................................(no files to check)Skipped
Check SPDX headers..............................................................Passed
Check root lazy imports.........................................................Passed
Check for spaces in all filenames...............................................Passed
Update Dockerfile dependency graph..............................................Passed
Check for forbidden imports.....................................................Passed
Prevent new 'torch.cuda' APIs call..............................................Passed
Validate configuration has default values and that each field has a docstring...Passed
Validate docker/versions.json matches Dockerfile................................(no files to check)Skipped
Check attention backend documentation is up to date.............................Passed
Check for boolean ops in with-statements........................................Passed
Suggestion......................................................................Passed
```

### CODE REVIEW

**Comment by jmamou** [comment]
@massif-01 

### Validation

Tested this pair (results below from a Xeon Platinum 8380 / vLLM CPU backend, but the validator change is platform-agnostic and applies equally on GPU):
- **Target:** `Qwen/Qwen2.5-Coder-14B-Instruct` (`vocab_size=152064`)
- **Draft:** `Qwen/Qwen2.5-Coder-0.5B-Instruct` (`vocab_size=151936`)

The padding gap is purely embedding-row padding — both tokenizers share the same 151,665 actual tokens (151,643 base + 22 added). With this PR's `allow_draft_model_vocab_padding=true`, SD runs cleanly and produces a substantial speedup on a code-completion workload (SPEED-Bench `low_entropy.code_completion`, 32 prompts, output_len=512):

| BS | k | TPOT (ms) | speedup vs baseline |
|---|---|---|---|
| 1 | baseline | 310 | 1.00× |
| 1 | 7 | 110 | **2.82×** |
| 4 | baseline | 690 | 1.00× |
| 4 | 7 | 348 | **1.98×** |
| 8 | baseline | 901 | 1.00× |
| 8 | 5 | 628 | **1.43×** |
| 8 | 7 | 609 | **1.48×** |

No correctness issues observed across the sweep. The flag is essential for this family because the equality check in `verify_equal_vocab_size_if_draft_model` blocks an otherwise-valid pair.

### Relationship to #38174

#38174 (universal heterogeneous tokenizer / TLI support) is the more general fix and would also cover this case (and others where the underlying tokenizers differ). This PR is complementary: a narrowly-scoped, low-risk opt-in for the common "same tokenizer, padded `vocab_size`" pattern that the Qwen2.5-Coder family exhibits today. Until #38174 lands, this flag unblocks a real-world deployment with a measurable speedup; once #38174 lands, this can be folded into the broader path.

It would be great to see this merged!

**Comment by massif-01** [comment]
> @massif-01
> 
> ### Validation
> Tested this pair (results below from a Xeon Platinum 8380 / vLLM CPU backend, but the validator change is platform-agnostic and applies equally on GPU):
> 
> * **Target:** `Qwen/Qwen2.5-Coder-14B-Instruct` (`vocab_size=152064`)
> * **Draft:** `Qwen/Qwen2.5-Coder-0.5B-Instruct` (`vocab_size=151936`)
> 
> The padding gap is purely embedding-row padding — both tokenizers share the same 151,665 actual tokens (151,643 base + 22 added). With this PR's `allow_draft_model_vocab_padding=true`, SD runs cleanly and produces a substantial speedup on a code-completion workload (SPEED-Bench `low_entropy.code_completion`, 32 prompts, output_len=512):
> 
> BS	k	TPOT (ms)	speedup vs baseline
> 1	baseline	310	1.00×
> 1	7	110	**2.82×**
> 4	baseline	690	1.00×
> 4	7	348	**1.98×**
> 8	baseline	901	1.00×
> 8	5	628	**1.43×**
> 8	7	609	**1.48×**
> No correctness issues observed across the sweep. The flag is essential for this family because the equality check in `verify_equal_vocab_size_if_draft_model` blocks an otherwise-valid pair.
> 
> ### Relationship to #38174
> #38174 (universal heterogeneous tokenizer / TLI support) is the more general fix and would also cover this case (and others where the underlying tokenizers differ). This PR is complementary: a narrowly-scoped, low-risk opt-in for the common "same tokenizer, padded `vocab_size`" pattern that the Qwen2.5-Coder family exhibits today. Until #38174 lands, this flag unblocks a real-world deployment with a measurable speedup; once #38174 lands, this can be folded into the broader path.
> 
> It would be great to see this merged!

Thanks a lot for the independent validation and detailed benchmark numbers!

This is exactly the narrow case this PR is trying to cover: same-tokenizer-family draft/target pairs where the actual token IDs are prefix-compatible, but the configured vocab sizes differ because the target has extra padded tail rows.

I also agree with your framing around #38174. TLI is the more general heterogeneous-tokenizer solution, while this PR is intended as a small opt-in path for the common padded-vocab case until the broader path lands.

**Comment by massif-01** [comment]
> This pull request has merge conflicts that must be resolved before it can be merged. Please rebase the PR, @massif-01.
> 
> https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

Thanks for the notice. I rebased this PR onto the latest `main` and force-pushed the updated branch.

GitHub now reports the PR as mergeable. The remaining `pre-run-check` failure appears to be the repository's label gate (`ready` / `verified`), not a merge conflict or code failure.

Local validation after the rebase:

- `git diff --check`: passed
- targeted pytest: `22 passed, 17 warnings`
- pre-commit on changed files: passed

Could a maintainer please review and add the appropriate label when convenient?
